component/grant-type/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/TokenExchangeGrantHandler.java

```diff
@@ -0,0 +1,272 @@
+            Date notBeforeTime = claimsSet.getNotBeforeTime();
+            Date issuedAtTime = claimsSet.getIssueTime();
+            Map<String, Object> customClaims = new HashMap<>(claimsSet.getClaims());
+
+            tokReqMsgCtx.addProperty(TokenExchangeConstants.EXPIRY_TIME, expirationTime);
+            long currentTimeInMillis = System.currentTimeMillis();
+            long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
+
+            if (StringUtils.isEmpty(jwtIssuer) || StringUtils.isEmpty(subject) || expirationTime == null ||
+                    audiences == null) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Mandatory fields(Issuer, Subject, Expiration time "
+                        + "or Audience) are empty in the given JWT");
+            }
+            identityProvider = TokenExchangeUtils.getIdPByIssuer(jwtIssuer, tenantDomain);
+            tokenEndPointAlias = TokenExchangeUtils.getTokenEndpointAlias(identityProvider, tenantDomain);
+            try {
+                if (signedJWT != null) {
+                    if (TokenExchangeUtils.validateSignature(signedJWT, identityProvider, tenantDomain)) {
+                        log.debug("Signature/MAC validated successfully.");
+                    } else {
+                        handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Signature or Message Authentication " +
+                                "invalid");
+                    }
+                }
+            } catch (JOSEException e) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Error when verifying signature");
+            }
+            TokenExchangeUtils.setAuthorizedUser(tokReqMsgCtx, identityProvider, subject);
+
+            log.debug("Subject(sub) found in JWT: " + subject);
+            log.debug(subject + " set as the Authorized User.");
+
+            tokReqMsgCtx.setScope(tokReqMsgCtx.getOauth2AccessTokenReqDTO().getScope());
+            if (StringUtils.isEmpty(tokenEndPointAlias)) {
+                handleException(OAuth2ErrorCodes.INVALID_REQUEST, "Token Endpoint alias of the local Identity " +
+                        "Provider has not been configured for " + identityProvider.getIdentityProviderName());
+            }
+
+            audienceFound = validateAudience(audiences, tokenEndPointAlias, requested_audience);
+            if (!audienceFound) {
+                handleException(TokenExchangeConstants.INVALID_TARGET, "Invalid audience values provided");
+            }
+            TokenExchangeUtils.checkExpirationTime(expirationTime, currentTimeInMillis, timeStampSkewMillis);
+            if (notBeforeTime != null) {
+                TokenExchangeUtils.checkNotBeforeTime(notBeforeTime, currentTimeInMillis, timeStampSkewMillis);
+            } else {
+                log.debug("Not Before Time(nbf) not found in JWT. Continuing Validation");
+            }
```

Comment:
can we move all the claim validation logic to a separate private method?

Reply:
Fixed